File: easyDiffractionApp/Gui/Logic/PyQmlProxy.js
class PyQmlProxy {

    // Properties

    get calculatorList() {
        return ["calculator1"]
    }

    get calculatorIndex() {
        return 0
    }

    get minimizerList() {
        return ["minimizer1"]
    }

    get minimizerIndex() {
        return 0
    }

    get amplitude() {
        return 3.5
    }

    get period() {
        return 2.0
    }

    get xShift() {
        return 0
    }

    get yShift() {
        return 0
    }

    get statusModelAsXml() {
        return "<root><item><label>Calculator</label><value>calculator1</value></item><item><label>Minimizer</label><value>minimizer1</value></item></root>"
    }

    get fitablesListAsXml() {
        return "<root><item><number>1</number><label>amplitude</label><value>3.5</value><unit></unit><error>0.0</error><fit>1</fit></item><item><number>2</number><label>period</label><value>3.141592653589793</value><unit></unit><error>0.0</error><fit>1</fit></item><item><number>3</number><label>x_shift</label><value>0.0</value><unit></unit><error>0.0</error><fit>1</fit></item><item><number>4</number><label>y_shift</label><value>0.0</value><unit></unit><error>0.0</error><fit>1</fit></item></root>"
    }

    get fitablesDict() {
        return {'amplitude': 3.5, 'period': 3.141592653589793, 'x_shift': 0.0, 'y_shift': 0.0}
    }

    get constraintsListAsXml() {
        return "<root><item><number>1</number><dependentName>amplitude</dependentName><relationalOperator>=</relationalOperator><value>1.0000</value><arithmeticOperator>*</arithmeticOperator><independentName>period</independentName><enabled>1</enabled></item><item><number>2</number><dependentName>amplitude</dependentName><relationalOperator>&lt;</relationalOperator><value>4.0000</value><arithmeticOperator></arithmeticOperator><independentName></independentName><enabled>1</enabled></item></root>"
    }

    get projectInfoAsJson() {
        return {"calculations":"experiments.cif","experiments":"experiments.cif","keywords":"sine, cosine, lmfit, bumps","modified":"18.09.2020, 09:24","name":"Example Project","samples":"samples.cif"}
    }

    get phasesXml() {
        //return "<root><item><label>Fe3O4</label><atoms><item><label>Fe1</label><type>Fe</type><x>0</x><y>0</y><z>0</z></item><item><label>Fe2</label><type>Fe</type><x>0.5</x><y>0</y><z>0</z></item><item><label>O</label><type>O</type><x>0.3421</x><y>0</y><z>0.5</z></item></atoms></item><item><label>CoO</label><atoms><item><label>Co</label><type>Co</type><x>0.5</x><y>0.25</y><z>0.5</z></item><item><label>O</label><type>O</type><x>0.75</x><y>0.75</y><z>0.75</z></item></atoms></item></root>"
        //return "<root><item><label>Sin_1</label><color>darkolivegreen</color><parameters><item><amplitude>3.2</amplitude><period>2.1</period></item></parameters></item><item><label>Sin_2</label><color>steelblue</color><parameters><item><amplitude>2.5</amplitude><period>2.7</period></item></parameters></item></root>"
        return "<root><item><label>Co2SiO4</label><color>darkolivegreen</color><crystal_system>orthorhombic</crystal_system><space_group_name>P n m a</space_group_name><space_group_setting>abc</space_group_setting><cell_length_a>10.28</cell_length_a><cell_length_b>10.28</cell_length_b><cell_length_c>10.28</cell_length_c><cell_angle_alpha>90.0</cell_angle_alpha><cell_angle_beta>90.0</cell_angle_beta><cell_angle_gamma>90.0</cell_angle_gamma><atoms><item><label>Co1</label><type>Co</type><color>coral</color><x>0.0</x><y>0.0</y><z>0.0</z><occupancy>1.0</occupancy><adp_type>Uiso</adp_type><adp_iso>0.004</adp_iso></item><item><label>Co2</label><type>Co</type><color>coral</color><x>0.279</x><y>0.279</y><z>0.279</z><occupancy>1.0</occupancy><adp_type>Uiso</adp_type><adp_iso>0.007</adp_iso></item><item><label>Si</label><type>Si</type><color>steelblue</color><x>0.094</x><y>0.094</y><z>0.094</z><occupancy>1.0</occupancy><adp_type>Uiso</adp_type><adp_iso>0.005</adp_iso></item><item><label>O1</label><type>O</type><color>darkolivegreen</color><x>0.091</x><y>0.091</y><z>0.091</z><occupancy>1.0</occupancy><adp_type>Uiso</adp_type><adp_iso>0.008</adp_iso></item><item><label>O2</label><type>O</type><color>darkolivegreen</color><x>0.448</x><y>0.448</y><z>0.448</z><occupancy>1.0</occupancy><adp_type>Uiso</adp_type><adp_iso>0.008</adp_iso></item><item><label>O3</label><type>O</type><color>darkolivegreen</color><x>0.164</x><y>0.164</y><z>0.164</z><occupancy>1.0</occupancy><adp_type>Uiso</adp_type><adp_iso>0.011</adp_iso></item></atoms></item><item><label>CoO</label><color>steelblue</color><crystal_system>cubic</crystal_system><space_group_name>F m -3 m</space_group_name><space_group_setting>1</space_group_setting><cell_length_a>3.02</cell_length_a><cell_length_b>3.02</cell_length_b><cell_length_c>3.02</cell_length_c><cell_angle_alpha>90.0</cell_angle_alpha><cell_angle_beta>90.0</cell_angle_beta><cell_angle_gamma>90.0</cell_angle_gamma><atoms><item><label>Co</label><type>Co</type><color>coral</color><x>0.0</x><y>0.0</y><z>0.0</z><occupancy>1.0</occupancy><adp_type>Uiso</adp_type><adp_iso>0.0</adp_iso></item><item><label>O</label><type>O</type><color>darkolivegreen</color><x>0.2471</x><y>0.4986</y><z>0.7497</z><occupancy>1.0</occupancy><adp_type>Uiso</adp_type><adp_iso>0.0</adp_iso></item></atoms></item></root>"
    }

    get phasesDict() {
        return [{"atoms":[{"adp_iso":0.004,"adp_type":"Uiso","color":"coral","label":"Co1","occupancy":1,"type":"Co","x":0,"y":0,"z":0},{"adp_iso":0.007,"adp_type":"Uiso","color":"coral","label":"Co2","occupancy":1,"type":"Co","x":0.279,"y":0.279,"z":0.279},{"adp_iso":0.005,"adp_type":"Uiso","color":"steelblue","label":"Si","occupancy":1,"type":"Si","x":0.094,"y":0.094,"z":0.094},{"adp_iso":0.008,"adp_type":"Uiso","color":"darkolivegreen","label":"O1","occupancy":1,"type":"O","x":0.091,"y":0.091,"z":0.091},{"adp_iso":0.008,"adp_type":"Uiso","color":"darkolivegreen","label":"O2","occupancy":1,"type":"O","x":0.448,"y":0.448,"z":0.448},{"adp_iso":0.011,"adp_type":"Uiso","color":"darkolivegreen","label":"O3","occupancy":1,"type":"O","x":0.164,"y":0.164,"z":0.164}],"cell_angle_alpha":90,"cell_angle_beta":90,"cell_angle_gamma":90,"cell_length_a":10.28,"cell_length_b":10.28,"cell_length_c":10.28,"color":"darkolivegreen","crystal_system":"orthorhombic","label":"Co2SiO4","space_group_name":"P n m a","space_group_setting":"abc"},{"atoms":[{"adp_iso":0,"adp_type":"Uiso","color":"coral","label":"Co","occupancy":1,"type":"Co","x":0,"y":0,"z":0},{"adp_iso":0,"adp_type":"Uiso","color":"darkolivegreen","label":"O","occupancy":1,"type":"O","x":0.2471,"y":0.4986,"z":0.7497}],"cell_angle_alpha":90,"cell_angle_beta":90,"cell_angle_gamma":90,"cell_length_a":3.02,"cell_length_b":3.02,"cell_length_c":3.02,"color":"steelblue","crystal_system":"cubic","label":"CoO","space_group_name":"F m -3 m","space_group_setting":"1"}]
    }

    get phasesCurrentIndex() {
        return 0
    }

    get parametersCurrentIndex() {
        return 0
    }
    // Functions

    addLowerMeasuredSeriesRef(series) {}

    addUpperMeasuredSeriesRef(series) {}

    setCalculatedSeriesRef(series) {}

    updateCalculatedData() {}

    generateMeasuredData() {}

    startFitting() {}

    editFitableValueByName(name, value) {}

    editFitableByIndexAndName(index, name, value) {}

    addConstraints(dependent_par_idx, operator, independent_par_idx) {}

    removeConstraintByIndex(index) {}

    toggleConstraintByIndex(index, enabled) {}

    editProjectInfoByKey(key, value) {}

    editPhase(p1, p2, p3) {}
}

File: easyExampleApp/Logic/PyQmlProxy.py
import json
from dicttoxml import dicttoxml
from distutils.util import strtobool

from PySide2.QtCore import QObject, Slot, Signal, Property
from PySide2.QtCharts import QtCharts

from easyCore import borg
from easyCore.Fitting.Fitting import Fitter
from easyCore.Fitting.Constraints import ObjConstraint, NumericConstraint
from easyCore.Utils.classTools import generatePath

from easyExampleLib.interface import InterfaceFactory
from easyExampleLib.model import Sin, DummySin

from easyExampleApp.Logic.QtDataStore import QtDataStore
from easyExampleApp.Logic.DisplayModels.DataModels import MeasuredDataModel, CalculatedDataModel


class PyQmlProxy(QObject):
    _borg = borg
    projectInfoChanged = Signal()
    modelChanged = Signal()
    constraintsChanged = Signal()
    calculatorChanged = Signal()
    minimizerChanged = Signal()
    statusChanged = Signal()
    phasesChanged = Signal()

    def __init__(self, parent=None):
        super().__init__(parent)
        self.phases = self.initPhases()
        self.project_info = self.initProjectInfo()
        self.interface = InterfaceFactory()
        self.model = Sin(self.interface)
        self.fitter = Fitter(self.model, self.interface.fit_func)
        self.dummy_source = DummySin()
        self.data = QtDataStore(self.dummy_source.x_data, self.dummy_source.y_data, self.dummy_source.sy_data, None)
        self._measured_data_model = MeasuredDataModel(self.data)
        self._calculated_data_model = CalculatedDataModel(self.data)
        self.updateCalculatedData()
        # when to emit status bar items cnahged
        self.calculatorChanged.connect(self.statusChanged)
        self.minimizerChanged.connect(self.statusChanged)

    # Data

    @Slot()
    def generateMeasuredData(self):
        self.dummy_source = DummySin()
        self.data = QtDataStore(self.dummy_source.x_data, self.dummy_source.y_data, self.dummy_source.sy_data, None)
        self._measured_data_model.updateData(self.data)

    @Slot()
    def updateCalculatedData(self):
        self.data.y_opt = self.interface.fit_func(self.data.x)
        self._calculated_data_model.updateData(self.data)
        self.modelChanged.emit()

    # Calculator

    @Property('QVariant', notify=calculatorChanged)
    def calculatorList(self):
        return self.interface.available_interfaces

    @Property(int, notify=calculatorChanged)
    def calculatorIndex(self):
        return self.calculatorList.index(self.interface.current_interface_name)

    @calculatorIndex.setter
    def setCalculator(self, index: int):
        self.model.switch_interface(self.calculatorList[index])
        self.updateCalculatedData()
        self.calculatorChanged.emit()

    # Minimizer

    @Property('QVariant', notify=minimizerChanged)
    def minimizerList(self):
        return self.fitter.available_engines

    @Property(int, notify=minimizerChanged)
    def minimizerIndex(self):
        return self.minimizerList.index(self.fitter.current_engine.name)

    @minimizerIndex.setter
    def setMinimizer(self, index: int):
        self.fitter.switch_engine(self.minimizerList[index])
        self.minimizerChanged.emit()

    # Calculator

    @Property('QVariant', notify=calculatorChanged)
    def calculatorList(self):
        return self.interface.available_interfaces

    @Property(int, notify=calculatorChanged)
    def calculatorIndex(self):
        return self.calculatorList.index(self.interface.current_interface_name)

    @calculatorIndex.setter
    def setCalculator(self, index: int):
        self.model.switch_interface(self.calculatorList[index])
        self.updateCalculatedData()
        self.calculatorChanged.emit()

    # Minimizer

    @Property('QVariant', notify=minimizerChanged)
    def minimizerList(self):
        return self.fitter.available_engines

    @Property(int, notify=minimizerChanged)
    def minimizerIndex(self):
        return self.minimizerList.index(self.fitter.current_engine.name)

    @minimizerIndex.setter
    def setMinimizer(self, index: int):
        self.fitter.switch_engine(self.minimizerList[index])
        self.minimizerChanged.emit()

    @Slot()
    def startFitting(self):
        result = self.fitter.fit(self.data.x, self.data.y, weights=self.data.sy)
        self.data.y_opt = result.y_calc
        self._calculated_data_model.updateData(self.data)
        self.modelChanged.emit()

    # Charts

    @Slot(QtCharts.QXYSeries)
    def addMeasuredSeriesRef(self, series):
        self._measured_data_model.addSeriesRef(series)

    @Slot(QtCharts.QXYSeries)
    def addLowerMeasuredSeriesRef(self, series):
        self._measured_data_model.addLowerSeriesRef(series)

    @Slot(QtCharts.QXYSeries)
    def addUpperMeasuredSeriesRef(self, series):
        self._measured_data_model.addUpperSeriesRef(series)

    @Slot(QtCharts.QXYSeries)
    def setCalculatedSeriesRef(self, series):
        self._calculated_data_model.setSeriesRef(series)

    # Undo/redo

    @Slot()
    def undo(self):
        self._borg.stack.undo()
        self.modelChanged.emit()

    @Slot(result=bool)
    def canUndo(self):
        return self._borg.stack.canUndo()

    @Slot()
    def redo(self):
        self._borg.stack.redo()
        self.modelChanged.emit()

    @Slot(result=bool)
    def canRedo(self):
        return self._borg.stack.canRedo()

    # Status

    @Property(str, notify=statusChanged)
    def statusModelAsXml(self):
        items = [ { "label": "Calculator", "value": self.interface.current_interface_name },
                  { "label": "Minimizer", "value": self.fitter.current_engine.name } ]
        xml = dicttoxml(items, attr_type=False)
        xml = xml.decode()
        return xml

    # Model parameters

    @Property(str, notify=modelChanged)
    def amplitude(self):
        return str(self.model.amplitude.raw_value)

    @amplitude.setter
    def setAmplitude(self, value: str):
        value = float(value)
        self.model.amplitude = value
        self.updateCalculatedData()

    @Property(str, notify=modelChanged)
    def period(self):
        return str(self.model.period.raw_value)

    @period.setter
    def setPeriod(self, value: str):
        value = float(value)
        self.model.period = value
        self.updateCalculatedData()

    @Property(str, notify=modelChanged)
    def xShift(self):
        return str(self.model.x_shift.raw_value)

    @xShift.setter
    def setXShift(self, value: str):
        value = float(value)
        self.model.x_shift = value
        self.updateCalculatedData()

    @Property(str, notify=modelChanged)
    def yShift(self):
        return str(self.model.y_shift.raw_value)

    @yShift.setter
    def setYShift(self, value: str):
        value = float(value)
        self.model.y_shift = value
        self.updateCalculatedData()

    # Models

    def fitablesList(self):
        fitables_list = []
        pars_id, pars_path = generatePath(self.model)
        for index, par_path in enumerate(pars_path):
            par = borg.map.get_item_by_key(pars_id[index])
            fitables_list.append(
                { "number": index + 1,
                  "label": par_path,
                  "value": par.raw_value,
                  "unit": str(par.unit).replace("dimensionless", ""),
                  "error": par.error,
                  "fit": int(not par.fixed) }
            )
        return fitables_list

    @Property(str, notify=modelChanged)
    def fitablesListAsXml(self):
        xml = dicttoxml(self.fitablesList(), attr_type=False)
        xml = xml.decode()
        return xml

    @Property('QVariant', notify=modelChanged)
    def fitablesDict(self):
        fitables_dict = {}
        for par in self.model.get_parameters():
            fitables_dict[par.name] = par.raw_value
        return fitables_dict

    @Slot(str, str)
    def editFitableValueByName(self, name, value):
        for par in self.model.get_parameters():
            if par.name == name:
                par.value = float(value)
                self.updateCalculatedData()

    @Slot(int, str, str)
    def editFitableByIndexAndName(self, index, name, value):
        #print("----", index, name, value)
        if index == -1: # TODO: Check why index is changed twice when name == "value"
            return
        par = self.model.get_parameters()[index]
        if name == "fit":
            par.fixed = not bool(strtobool(value))
        elif name == "value":
            par.value = float(value)
            self.updateCalculatedData()
        else:
            print(f"Unsupported name '{name}'")

    # Constraints

    @Slot(int, str, str, str, int)
    def addConstraint(self, dependent_par_idx, relational_operator,
                      value, arithmetic_operator, independent_par_idx):
        if dependent_par_idx == -1 or value == "":
            print("Failed to add constraint: Unsupported type")
            return
        #if independent_par_idx == -1:
        #    print(f"Add constraint: {self.fitablesList()[dependent_par_idx]['label']}{relational_operator}{value}")
        #else:
        #    print(f"Add constraint: {self.fitablesList()[dependent_par_idx]['label']}{relational_operator}{value}{arithmetic_operator}{self.fitablesList()[independent_par_idx]['label']}")
        pars = self.model.get_parameters()
        if arithmetic_operator != "" and independent_par_idx > -1:
            c = ObjConstraint(pars[dependent_par_idx],
                              str(float(value)) + arithmetic_operator,
                              pars[independent_par_idx])
        elif arithmetic_operator == "" and independent_par_idx == -1:
            c = NumericConstraint(pars[dependent_par_idx],
                                  relational_operator.replace("=", "=="),
                                  float(value))
        else:
            print("Failed to add constraint: Unsupported type")
            return
        #print(c)
        c()
        self.fitter.add_fit_constraint(c)
        self.constraintsChanged.emit()
        self.updateCalculatedData()

    def constraintsList(self):
        constraint_list = []
        for index, constraint in enumerate(self.fitter.fit_constraints()):
            if type(constraint) is ObjConstraint:
                independent_name = constraint.get_obj(constraint.independent_obj_ids).name
                relational_operator = "="
                value = float(constraint.operator[:-1])
                arithmetic_operator = constraint.operator[-1]
            elif type(constraint) is NumericConstraint:
                independent_name = ""
                relational_operator = constraint.operator.replace("==", "=")
                value = constraint.value
                arithmetic_operator = ""
            else:
                print(f"Failed to get constraint: Unsupported type {type(constraint)}")
                return
            number = index + 1
            dependent_name = constraint.get_obj(constraint.dependent_obj_ids).name
            enabled = int(constraint.enabled)
            constraint_list.append(
                { "number": number,
                  "dependentName": dependent_name,
                  "relationalOperator": relational_operator,
                  "value": value,
                  "arithmeticOperator": arithmetic_operator,
                  "independentName": independent_name,
                  "enabled": enabled }
            )
        return constraint_list

    @Property(str, notify=constraintsChanged)
    def constraintsListAsXml(self):
        xml = dicttoxml(self.constraintsList(), attr_type=False)
        xml = xml.decode()
        return xml

    @Slot(int)
    def removeConstraintByIndex(self, index: int):
        self.fitter.remove_fit_constraint(index)
        self.constraintsChanged.emit()

    @Slot(int, str)
    def toggleConstraintByIndex(self, index, enabled):
        constraint = self.fitter.fit_constraints()[index]
        constraint.enabled = bool(strtobool(enabled))
        self.constraintsChanged.emit()
        self.updateCalculatedData()

    # App project info

    def initProjectInfo(self):
        return dict(name = "Example Project",
                    keywords = "sine, cosine, lmfit, bumps",
                    samples = "samples.json",
                    experiments = "experiments.json",
                    calculations = "calculation.json",
                    modified = "18.09.2020, 09:24")

    @Property('QVariant', notify=projectInfoChanged)
    def projectInfoAsJson(self):
        return self.project_info

    @projectInfoAsJson.setter
    def setProjectInfoAsJson(self, json_str):
        self.project_info = json.loads(json_str)
        self.projectInfoChanged.emit()

    @Slot(str, str)
    def editProjectInfoByKey(self, key, value):
        self.project_info[key] = value
        self.projectInfoChanged.emit()

    # Test phases dict

    def initPhases(self):
        phases = [dict(label = "Sin_1",
                     color = "darkolivegreen",
                     parameters = [dict(amplitude = 3.2, period = 2.1)]
                ),
                dict(label = "Sin_2",
                     color = "steelblue",
                     parameters = [dict(amplitude = 2.5, period = 2.7)]
                )
               ]
        return phases

    @Property('QVariant', notify=phasesChanged)
    def phasesDict(self):
        return self.phases

    @phasesDict.setter
    def setPhasesDict(self, json_str):
        self.phases = json.loads(json_str)
        self.phasesChanged.emit()

    @Property(str, notify=phasesChanged)
    def phasesXml(self):
        xml = dicttoxml(self.phases, attr_type=False)
        xml = xml.decode()
        return xml

    @Slot(int, str, str)
    def editPhase(self, phase_index, parameter_name, new_value):
        #print("----", phase_index, parameter_name, new_value)
        self.phases[phase_index][parameter_name] = new_value
        self.phasesChanged.emit()

    @Slot(int, int, str, str)
    def editPhaseParameter(self, phase_index, parameter_index, parameter_name, new_value):
        #print("----", phase_index, parameter_index, parameter_name, new_value)
        self.phases[phase_index]['parameters'][parameter_index][parameter_name] = new_value
        self.phasesChanged.emit()
